File: native/computer-use/Sources/OperonComputerUseServer/PeerAuth.swift
import Foundation
import Security
#if canImport(Darwin)
import Darwin
#endif

// A `<sys/un.h>` constant the Swift Darwin overlay does not always export; defined here from the
// header value as a fallback.
private let kSOL_LOCAL: Int32 = 0
private let kLOCAL_PEERTOKEN: Int32 = 0x006

/// codesign peer verification (server side): reads the codesign Team ID from the audit token of
/// the peer on the incoming socket and compares it to this process's own. Same policy as
/// `packages/browser-use/peer-auth.ts`:
///
/// - `OPERON_REQUIRE_SIGNED_PEER` unset → no check.
/// - no Team ID on self (dev / adhoc) → nothing to enforce against → warn and allow
///   (**never lock a developer out**).
/// - Team ID present on self (a signed release build) → enforce: reject any peer whose Team is
///   not on the allowlist.
///
/// The happy path can only be verified on a notarised build. This layers on top of the phase 1
/// startup token, giving two gates.
enum PeerAuth {
    private static let enableEnvKey = "OPERON_REQUIRE_SIGNED_PEER"
    private static let extraTeamsEnvKey = "OPERON_ALLOWED_PEER_TEAMS"

    private static let enabled: Bool =
        ProcessInfo.processInfo.environment[enableEnvKey]?.isEmpty == false

    /// The allowlist is this process's own Team ID plus anything extra configured. With no Team ID
    /// of our own (adhoc), this is nil and nothing can be enforced.
    private static let allowedTeams: Set<String>? = {
        guard let selfTeam = teamIdentifier(for: copySelfCode()) else { return nil }
        let extra = (ProcessInfo.processInfo.environment[extraTeamsEnvKey] ?? "")
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
        return Set([selfTeam] + extra)
    }()

    static func authorize(fd: Int32) -> Bool {
        guard enabled else { return true }
        guard let teams = allowedTeams else {
            warnOnce("unsigned-self",
                     "\(enableEnvKey) set but this build is unsigned (no Team ID); allowing (cannot enforce)")
            return true
        }
        guard let peerTeam = peerTeamIdentifier(fd: fd) else {
            log("rejected connection: peer has no team id")
            return false
        }
        if teams.contains(peerTeam) { return true }
        log("rejected connection: peer team \(peerTeam) not in allowlist")
        return false
    }

    // MARK: - SecCode

    private static func copySelfCode() -> SecCode? {
        var code: SecCode?
        guard SecCodeCopySelf(SecCSFlags(), &code) == errSecSuccess else { return nil }
        return code
    }

    private static func peerTeamIdentifier(fd: Int32) -> String? {
        var token = audit_token_t(val: (0, 0, 0, 0, 0, 0, 0, 0))
        var len = socklen_t(MemoryLayout<audit_token_t>.size)
        let rc = withUnsafeMutablePointer(to: &token) { ptr in
            getsockopt(fd, kSOL_LOCAL, kLOCAL_PEERTOKEN, ptr, &len)
        }
        guard rc == 0, Int(len) == MemoryLayout<audit_token_t>.size else { return nil }
        let data = withUnsafeBytes(of: token) { Data($0) } as CFData
        let attrs = [kSecGuestAttributeAudit: data] as CFDictionary
        var code: SecCode?
        guard SecCodeCopyGuestWithAttributes(nil, attrs, SecCSFlags(), &code) == errSecSuccess,
              let code
        else { return nil }
        return teamIdentifier(for: code)
    }

    private static func teamIdentifier(for code: SecCode?) -> String? {
        guard let code else { return nil }
        var staticCode: SecStaticCode?
        guard SecCodeCopyStaticCode(code, SecCSFlags(), &staticCode) == errSecSuccess,
              let staticCode
        else { return nil }
        var info: CFDictionary?
        let flags = SecCSFlags(rawValue: kSecCSSigningInformation)
        guard SecCodeCopySigningInformation(staticCode, flags, &info) == errSecSuccess,
              let dict = info as? [String: Any]
        else { return nil }
        return dict[kSecCodeInfoTeamIdentifier as String] as? String
    }

    // MARK: - logging

    private static let warnLock = NSLock()
    // Guarded by warnLock. Swift 6 strict concurrency cannot see that, so say it explicitly.
    nonisolated(unsafe) private static var warned = Set<String>()
    private static func warnOnce(_ key: String, _ message: String) {
        warnLock.lock()
        defer { warnLock.unlock() }
        if warned.contains(key) { return }
        warned.insert(key)
        log(message)
    }

    private static func log(_ message: String) {
        FileHandle.standardError.write(Data("[operon-cu peer-auth] \(message)\n".utf8))
    }
}

File: packages/browser-use/peer-auth.ts
// Policy layer for code-signature peer verification, server side.
//
// It decides whether a socket connecting into IabBackend or ChromeNativeHost
// should be admitted: read the peer's code-signing identity from its audit token
// through the peer-auth native addon, and compare it against this process's own
// Team ID. Obtaining the identity lives in
// `native/peer-auth/src/peer_auth.mm`; this file is policy and graceful
// degradation only.
//
// Verification is one-way and server-side: no secret is ever sent to the peer.
//
// Off by default, adaptive, and never able to lock a developer out:
//   - `OPERON_REQUIRE_SIGNED_PEER` unset: no check at all.
//   - addon missing, or not macOS: warn once and admit.
//   - this process has no Team ID (a dev or adhoc build, `TeamIdentifier=not set`):
//     nothing to enforce against, so warn and admit.
//   - this process has a Team ID (a signed, notarised release): enforce, and
//     refuse a peer whose team is not on the allowlist.
//
// The positive path, accepting our own team and refusing others, can only be
// verified on a notarised build: everything on a development machine is adhoc.

import { createRequire } from "node:module";
import path from "node:path";
import { existsSync } from "node:fs";
import { fileURLToPath } from "node:url";
import type { Socket } from "node:net";

interface PeerIdentity {
  teamIdentifier: string | null;
  signingIdentifier: string | null;
}

interface PeerAuthAddon {
  peerAuditIdentity(fd: number): PeerIdentity;
  selfAuditIdentity(): PeerIdentity;
}

const ENABLE_ENV = "OPERON_REQUIRE_SIGNED_PEER";
/** Additional Team IDs to admit, comma separated. Rarely needed; it exists for
 *  builds signed under more than one identity. */
const EXTRA_TEAMS_ENV = "OPERON_ALLOWED_PEER_TEAMS";

const require = createRequire(import.meta.url);

const warned = new Set<string>();
function warnOnce(key: string, message: string): void {
  if (warned.has(key)) return;
  warned.add(key);
  console.warn(`[peer-auth] ${message}`);
}

function loadAddon(): PeerAuthAddon | undefined {
  if (process.platform !== "darwin") return undefined;
  const here = path.dirname(fileURLToPath(import.meta.url)); // packages/browser-use
  // resourcesPath is Electron-only and invisible to Node's types, hence the cast.
  const resourcesPath = (process as unknown as { resourcesPath?: string }).resourcesPath;
  const candidates = [
    // Packaged: ships inside resources alongside operon-runtime.
    ...(resourcesPath
      ? [path.join(resourcesPath, "operon-runtime", "operon-peer-auth.node")]
      : []),
    // Build output inside the repository.
    path.join(here, "..", "..", "dist-operon-runtime", "operon-peer-auth.node"),
    path.join(here, "..", "..", "native", "peer-auth", "build", "Release", "operon_peer_auth.node"),
  ];
  for (const candidate of candidates) {
    if (!existsSync(candidate)) continue;
    try {
      return require(candidate) as PeerAuthAddon;
    } catch (error) {
      warnOnce(`load:${candidate}`, `failed to load ${candidate}: ${String(error)}`);
    }
  }
  return undefined;
}

let addonCache: PeerAuthAddon | undefined | null = null; // null means not yet attempted
function addon(): PeerAuthAddon | undefined {
  if (addonCache === null) addonCache = loadAddon();
  return addonCache ?? undefined;
}

let allowedTeamsCache: Set<string> | undefined | null = null; // null means not yet computed
/** The allowlist is this process's own Team ID plus any configured extras.
 *  Returns undefined when this process has no Team ID, meaning nothing can be
 *  enforced. */
function allowedTeams(): Set<string> | undefined {
  if (allowedTeamsCache !== null) return allowedTeamsCache ?? undefined;
  const mod = addon();
  const selfTeam = mod?.selfAuditIdentity().teamIdentifier ?? null;
  if (!selfTeam) {
    allowedTeamsCache = undefined;
    return undefined;
  }
  const extra = (process.env[EXTRA_TEAMS_ENV] ?? "")
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);
  allowedTeamsCache = new Set([selfTeam, ...extra]);
  return allowedTeamsCache;
}

/** The unix socket's underlying fd (`_handle.fd`), or undefined when unavailable. */
export function socketFd(socket: Socket): number | undefined {
  const handle = (socket as unknown as { _handle?: { fd?: number } })._handle;
  const fd = handle?.fd;
  return typeof fd === "number" && fd >= 0 ? fd : undefined;
}

/**
 * Whether to admit a connection; see the policy at the top of this file.
 *
 * It fails open: anything undecidable, whether the switch is off, the addon is
 * missing, this process is unsigned, or the fd cannot be read, is admitted. This
 * is defence in depth layered on top of 0700 and 0600 permissions, and a false
 * negative must not break the product.
 */
export function authorizePeer(socket: Socket): boolean {
  if (!process.env[ENABLE_ENV]) return true;

  const mod = addon();
  if (!mod) {
    warnOnce("no-addon", `${ENABLE_ENV} set but peer-auth addon unavailable; allowing (cannot enforce)`);
    return true;
  }

  const teams = allowedTeams();
  if (!teams) {
    warnOnce("unsigned-self", `${ENABLE_ENV} set but this build is unsigned (no Team ID); allowing (cannot enforce)`);
    return true;
  }

  const fd = socketFd(socket);
  if (fd == null) {
    warnOnce("no-fd", "could not read socket fd; allowing (cannot enforce)");
    return true;
  }

  const peerTeam = mod.peerAuditIdentity(fd).teamIdentifier;
  if (peerTeam && teams.has(peerTeam)) return true;
  console.warn(`[peer-auth] rejected connection: peer team ${peerTeam ?? "<none>"} not in allowlist`);
  return false;
}
